Shark Cage/createACLinNewDesktop/createACLinNewDesktop/util.cpp

```diff
@@ -133,75 +133,106 @@ namespace util {
 			tokenDeconstructed.TokenDefaultDacl,
 			tokenDeconstructed.TokenSource
 		);
+
+		//disable needed privileges
+		changeTokenCreationPrivilege(false);
+
+		if(!NT_SUCCESS(status)) {
+			if (NT_ERROR(status)) {
+				wprintf(L"  Cannot construct a token\n");
+				return false;
+			}
+		}
+
+		token = newToken;
+		return true;
 	}
 
 }
+
+//private functions
+
 ULONG getCurrentSessionID() {
-	//how to get correct sessionID even when I am under LOCAL System?
+	DWORD count = 0;
+	PWTS_SESSION_INFO  info;
+	WTSEnumerateSessions(WTS_CURRENT_SERVER_HANDLE, 0,1,&info, &count);
+	for (size_t i = 0; i < count; i++)
+	{
+		if (lstrcmp(info[i].pWinStationName, L"Console") == 0)
+		{
+			return info[i].SessionId;
+		}
+
+	}
+	return 0;
 }
 
+//deconstructs the template token
 void deconstructToken(tokenStructures &tokenDeconstructed, HANDLE &userToken) {
-	DWORD bufferSize = 0;
 
+	DWORD bufferSize = 0;
 	GetTokenInformation(userToken, TokenType, NULL, 0, &bufferSize);
-	GetTokenInformation(userToken, TokenType, (LPVOID) tokenDeconstructed.TokenType, bufferSize, &bufferSize);
+	SetLastError(0);
+	GetTokenInformation(userToken, TokenType, (LPVOID) &tokenDeconstructed.TokenType, bufferSize, &bufferSize);
 
 	bufferSize = 0;
 	GetTokenInformation(userToken, TokenUser, NULL, 0, &bufferSize);
+	SetLastError(0);
 	tokenDeconstructed.TokenUser = (PTOKEN_USER) new BYTE[bufferSize];

```

Comment:
all `new` calls in this function never get free'd, are they? shoud there be a `cleanToken` function?

Reply:
yes, indeed... I will be adding it in the next iteration (reasoning explained in a different comment)

Reply:
Fixed 😄